cupyx/scipy/ndimage/filters.py

```diff
@@ -969,381 +993,147 @@ def _get_rank_kernel(filter_size, rank, mode, wshape, origins, cval, int_type):
-    output array.
-
-    weights and structure can be given as None (structure defaults to None) in
-    which case they are not passed to the kernel at all. If the output is given
-    as None then it will be allocated in this function.
-
-    This function deals with making sure that the weights and structure are
-    contiguous and float64 (or bool for weights that are footprints)*, that the
-    output is allocated and appriopately shaped. This also deals with the
-    situation that the input and output arrays overlap in memory.
-
-    * weights is always cast to float64 or bool in order to get an output
-    compatible with SciPy, though float32 might be sufficient when input dtype
-    is low precision. If weights_dtype is passed as weights.dtype then no
-    dtype conversion will occur. The input and output are never converted.
+    Args:
+        input (cupy.ndarray): The input array.
+        function (cupy.RawKernel): The kernel to apply along each axis.
+        filter_size (int): Length of the filter.
+        axis (int): The axis of input along which to calculate. Default is -1.
+        output (cupy.ndarray, dtype or None): The array in which to place the
+            output. Default is is same dtype as the input.
+        mode (str): The array borders are handled according to the given mode
+            (``'reflect'``, ``'constant'``, ``'nearest'``, ``'mirror'``,
+            ``'wrap'``). Default is ``'reflect'``.
+        cval (scalar): Value to fill past edges of input if mode is
+            ``'constant'``. Default is ``0.0``.
+        origin (int): The origin parameter controls the placement of the
+            filter, relative to the center of the current element of the
+            input. Default is ``0``.
+    Returns:
+        cupy.ndarray: The result of the filtering.
+
+    .. notes::
+        The provided function (as a RawKernel) must have the following
+        signature. Unlike most functions, this should not utilize
+        `blockDim`/`blockIdx`/`threadIdx`.
+
+            __global__ void func(double *input_line, ptrdiff_t input_length,
+                                 double *output_line, ptrdiff_t output_length)
+
+    .. seealso:: :func:`scipy.ndimage.generic_filter1d`
     """
-    args = [input]
-    if weights is not None:
-        weights = cupy.ascontiguousarray(weights, weights_dtype)
-        args.append(weights)
-    if structure is not None:
-        structure = cupy.ascontiguousarray(structure, structure_dtype)
-        args.append(structure)
+    # This filter is very different than all other filters (including
+    # generic_filter and all 1d filters) and it has a customized solution.
+    # It is also likely fairly terrible, but only so much can be done when
+    # matching the scipy interface of having the sub-kernel work on entire
+    # lines of data.
+    from cupyx.scipy.ndimage._filters_generic import _get_generic_filter1d
```

Comment:
I think the CuPy style will require moving this to the top of the file and using

`from cupyx.scipy.ndimage import _filters_generic`
and then calling via `_filters_generic._get_generic_filter1d`.